Pass additional data for correct TextIndicator styling.

megangardner opened the pull request:
#### cc61ebcd1d9dffbc763dfb9d72050db3ea27b594
<pre>
Pass additional data for correct TextIndicator styling.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274505">https://bugs.webkit.org/show_bug.cgi?id=274505</a>
<a href="https://rdar.apple.com/128475613">rdar://128475613</a>

Reviewed by Aditya Keerthi.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextIndicatorStyleForID:withData:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _enableTextIndicatorStylingForElementWithID:]):
(-[WKWebView _addTextIndicatorStyleForID:withStyleType:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/TextIndicatorStyle.h:
* Source/WebKit/UIProcess/WKSTextStyleManager.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WKTextIndicatorStyleManager.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForUUID):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/279165@main">https://commits.webkit.org/279165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/892f93d4e683593fe0133ecd0fe62775a053f85f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52633 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31970 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5159 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3356 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3122 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42762 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2160 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54731 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/45479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23862 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/52477 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1515 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2928 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57503 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27770 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/2928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50153 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28994 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45620 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49424 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11499 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29908 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7730 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes reviewed (commented):
Can you use an ObjectIdentifier for remainingRangeUUID instead of UUID ?

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1789,15 +1789,15 @@ - (void)_storeAppHighlight:(const WebCore::AppHighlight&)highlight
 {
 #if PLATFORM(IOS_FAMILY)
-    [_contentView addTextIndicatorStyleForID:nsUUID withStyleType:styleType];
+    [_contentView addTextIndicatorStyleForID:nsUUID withStyleType:(WKTextIndicatorStyleType)data.style];
```
I don't think this cast is that safe since they are different types

pxlcoder replied:
Right, is there a reason why we need both enums? (maybe it was for some Swift-interop?)

But if we do need both, we should have a method to convert between them, similar to `static inline OptionSet<WebKit::FindOptions> toFindOptions(_WKFindOptions wkFindOptions)`.

megangardner replied:
Yes, we need both, I will make a converter.

rr-codes commented on Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
```diff
@@ -69,4 +69,11 @@ header: "TextIndicatorStyle.h"
+header: "TextIndicatorStyle.h"
+[CustomHeader] struct WebKit::TextIndicatorStyleData {
+    WebKit::TextIndicatorStyle style;
+    WTF::UUID remainingRangeUUID;
```
(idk if my previous comment actually showed up, but I think this should be an `ObjectIdentifier` if possible)

megangardner replied:
Sorry, not possible today.

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
```diff
@@ -25,6 +25,8 @@
 
 #pragma once
 
+
```
my greatest enemy: newlines 😛

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -89,6 +89,7 @@
 #import <WebCore/ScrollbarTheme.h>
 #import <WebCore/Settings.h>
 #import <WebCore/SharedBuffer.h>
+#import <WebCore/TextIndicator.h>
```
unified sources I'm guessing?

megangardner replied:
Yes :-/

rr-codes commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
```diff
@@ -109,6 +109,7 @@ class UnifiedTextReplacementController final {
 
     using TextIndicatorCharacterRange = std::pair<WTF::UUID, WebCore::CharacterRange>;
     Vector<std::pair<WTF::UUID, Vector<TextIndicatorCharacterRange>>> m_textIndicatorCharacterRangesForSessions;
+    Vector<std::pair<WTF::UUID, std::pair<WTF::UUID, WebCore::SimpleRange>>> m_remainingRangesForSessions;
```
can this be a `HashMap<UUID, HashMap<UUID, SimpleRange>>`?

megangardner replied:
Not today.

rr-codes commented on Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
```diff
@@ -25,6 +25,8 @@
 
 #pragma once
 
+
```
I implicitly meant that these extra newlines should be removed but I guess that wasn't quite obvious

megangardner replied:
I got some of them, but I guess not all of them?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3814,6 +3814,19 @@ - (void)_setFindDelegate:(id<_WKFindDelegate>)findDelegate
 
+static inline WKTextIndicatorStyleType toWKTextIndicatorStyleType(WebKit::TextIndicatorStyle style)
+{
+
```
Extra newline.

megangardner committed:
Pass additional data for correct TextIndicator styling.
https://bugs.webkit.org/show_bug.cgi?id=274505
rdar://128475613

Reviewed by Aditya Keerthi.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextIndicatorStyleForID:withData:]):
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _enableTextIndicatorStylingForElementWithID:]):
(-[WKWebView _addTextIndicatorStyleForID:withStyleType:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/Cocoa/WKTextIndicatorStyleType.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextIndicatorStyleForID):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/TextIndicatorStyle.h:
* Source/WebKit/UIProcess/WKSTextStyleManager.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setTextIndicatorFromFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WKTextIndicatorStyleManager.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForUUID):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextIndicatorStyleForID):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/279165@main
